primitives/manta/src/assets.rs

```diff
@@ -14,178 +14,244 @@
+    type Location;
+}
+
+/// Location Type
+pub type Location<T> = <T as LocationType>::Location;
 
-/// The minimal interface of asset metadata
-pub trait AssetMetadata {
-    /// Returns the minimum balance to hold this asset
-    fn min_balance(&self) -> Balance;
+/// Asset Metadata
+pub trait AssetMetadata: BalanceType {
+    /// Returns the minimum balance to hold this asset.
+    fn min_balance(&self) -> &Self::Balance;
 
-    /// Returns a boolean value indicating whether this asset needs an existential deposit
+    /// Returns a boolean value indicating whether this asset needs an existential deposit.
     fn is_sufficient(&self) -> bool;
 }
 
-/// The registrar trait: defines the interface of creating an asset in the asset implementation
+/// Asset Registry
+///
+/// The registry trait: defines the interface of creating an asset in the asset implementation
 /// layer. We may revisit this interface design (e.g. add change asset interface). However, change
 /// in StorageMetadata should be rare.
-pub trait AssetRegistrar<C: frame_system::Config, T: AssetConfig<C>> {
-    /// Create an new asset.
+pub trait AssetRegistry: AssetIdType + BalanceType {
+    /// Metadata Type
+    type Metadata;
+
+    /// Error Type
+    type Error;
+
+    /// Creates an new asset.
     ///
     /// * `asset_id`: the asset id to be created
-    /// * `min_balance`: the minimum balance to hold this asset
     /// * `metadata`: the metadata that the implementation layer stores
+    /// * `min_balance`: the minimum balance to hold this asset
     /// * `is_sufficient`: whether this asset can be used as reserve asset,
     ///     to the first approximation. More specifically, Whether a non-zero balance of this asset
     ///     is deposit of sufficient value to account for the state bloat associated with its
     ///     balance storage. If set to `true`, then non-zero balances may be stored without a
     ///     `consumer` reference (and thus an ED in the Balances pallet or whatever else is used to
     ///     control user-account state growth).
     fn create_asset(
-        asset_id: AssetId,
-        min_balance: Balance,
-        metadata: T::StorageMetadata,
+        asset_id: Self::AssetId,
+        metadata: Self::Metadata,
+        min_balance: Self::Balance,
         is_sufficient: bool,
-    ) -> DispatchResult;
+    ) -> Result<(), Self::Error>;
 
     /// Update asset metadata by `AssetId`.
     ///
     /// * `asset_id`: the asset id to be created.
     /// * `metadata`: the metadata that the implementation layer stores.
-    fn update_asset_metadata(asset_id: AssetId, metadata: T::StorageMetadata) -> DispatchResult;
+    fn update_asset_metadata(
+        asset_id: &Self::AssetId,
+        metadata: Self::Metadata,
+    ) -> Result<(), Self::Error>;
 }
 
-pub trait AssetConfig<C>: 'static + Eq + Clone
+/// Asset Configuration
+pub trait AssetConfig<C>: AssetIdType + BalanceType + LocationType
 where
-    C: frame_system::Config,
+    C: Config,
```

Comment:
Do we really need frame_system::Config trait bound? Can we just use `<AccountId>` type parameter?

Reply:
I'm actually not sure we can get away with just `AccountId` generic (or maybe i'm misunderstanding you)
You can log and give it a try separately, i feel like it will take a while to disentangle the code if it can be implemented that way.